Haskell symbolic lookup rule

ana-pantilie opened the pull request:
Fixes https://github.com/kframework/kore/issues/2247

ana-pantilie committed:
Add symbolic lookup simplification rule

ana-pantilie committed:
Add map-symbolic-tests-haskell

ehildenb commented:
@ana-pantilie did you make changes to the source files that `gen-tests.sh` uses? If so, please commit those changes too.

You may also just consider re-using the existing `map-symbolic-tests`, and changing it to using the Haskell backend instead of the Java backend, and disabling any tests which don't work now.

ana-pantilie commented:
@ehildenb Thanks for taking a look! I didn't make any changes to `gen-tests.sh`, but I had to make some changes to the tests themselves (I moved the old expected output to the RHS of the claim, since we expect these specs to be proved). I removed all the tests which use `update`, since we don't have a symbolic rule for it yet. A few of the tests I added don't work, but I haven't investigated much yet to see what might be the problem (they're not related to the rule I added, but I want to discuss them first), so this is still a draft.

andreiburdusa committed:
Rule and "assignment-" tests

ehildenb commented:
Well I mean more specifically that these tests are generated using `gen-tests.sh`, and the source files (like `assignment` and `lookup` files). So just make sure those source files for generating the tests are committed, similar to what @andreiburdusa has done here: https://github.com/kframework/k/pull/1662/files

andreiburdusa committed:
Fix rules

andreiburdusa committed:
Some tests

andreiburdusa committed:
Some tests

andreiburdusa committed:
Fix tests

ttuegel committed:
map-symbolic-tests-haskell: Update assignment specs

ttuegel committed:
domains.md: Add a rule for double Map update

ttuegel committed:
domains.md: Add a rule for swapping double Map updates

ttuegel committed:
map-symbolic-tests-haskell/gen-tests.sh: Update specs unconditionally

andreiburdusa committed:
Generate tests for lookup, inkeys, and remove

andreiburdusa committed:
Merge remote-tracking branch 'upstream/master' into haskell-symbolic-MAP.update

ana-pantilie committed:
Generate tests; use old RHS

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup-7-spec.k.out:
```diff
@@ -0,0 +1 @@
+#Top
```
On the Java backend, the result is
```
<k>
  lookupResult ( false )
</k>
```

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys:
```diff
@@ -0,0 +1,9 @@
+claim <k> inkeys ( X:MyId in_keys ( .Map  ) ) => . </k>
+claim <k> inkeys ( Y:MyId in_keys ( M:Map ) ) => . </k> requires Y in_keys(M)
+claim <k> inkeys ( Y:MyId in_keys ( M:Map ) ) => . </k> requires notBool Y in_keys(M)
```
These tests are missing from the Java backend test suite, even though they exist in the `inkeys` file.

ttuegel replied:
We don't expect these to pass until we finish kframework/kore#2249.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-2-spec.k.out:
```diff
@@ -0,0 +1,9 @@
+  <k>
+    inkeysResult ( Y:MyId in_keys ( M ) ) ~> .
```
Should this be `inkeysResult ( true )`?

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-3-spec.k.out:
```diff
@@ -0,0 +1,9 @@
+  <k>
+    inkeysResult ( Y:MyId in_keys ( M ) ) ~> .
```
Should this be `inkeysResult ( false )`?

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-6-spec.k.out:
```diff
@@ -0,0 +1,16 @@
+  <k>
+    inkeysResult ( Y:MyId in_keys ( M
+    ( X:MyId |-> 0 ) ) ) ~> .
```
This is the expected output on the Java backend as well, but I don't understand what this really tests. Maybe it would make sense if the expected output is `inkeysResult ( false )`, and we add `X =/=K Y` to the requires? I tried this out and it fails, but it's just an idea.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-8-spec.k.out:
```diff
@@ -0,0 +1,23 @@
+  } )
+#And
+  <k>
+    inkeysResult ( Y:MyId in_keys ( M
+    X:MyId |-> 0
+    ( Z:MyId |-> 1 ) ) ) ~> .
```
This is the expected output on the Java backend as well, I don't understand what this tests.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys:
```diff
@@ -0,0 +1,9 @@
+claim <k> inkeys ( Y:MyId in_keys ( M:Map X:MyId |-> 0 ) ) => . </k> requires notBool Y in_keys(M)
+claim <k> inkeys ( X:MyId in_keys ( Z:MyId |-> 1 M:Map X      |-> 0 ) ) => . </k>
+claim <k> inkeys ( Y:MyId in_keys ( Z:MyId |-> 1 M:Map X:MyId |-> 0 ) ) => . </k>
+claim <k> inkeys ( Y:MyId in_keys ( Z:MyId |-> 1 M:Map X:MyId |-> 0 ) ) => . </k> requires notBool Y in_keys(M)
```
I can't find this test either.

andreiburdusa committed:
Add remove-8 and remove-12

andreiburdusa committed:
Remove lookup tests since Ana is working on them

andreiburdusa committed:
Add rules to make tests pass

andreiburdusa committed:
Merge remote-tracking branch 'upstream/master' into haskell-symbolic-MAP.update

ana-pantilie committed:
Use fail; regenerate all tests with expected RHS; move failing to failing

ana-pantilie committed:
Merge remote-tracking branch 'andrei-burdusa/haskell-symbolic-MAP.update' into haskell-symbolic-lookup

ana-pantilie committed:
Revert change

ana-pantilie committed:
Add previously failing tests

ana-pantilie committed:
Add simplification rule; new test results

ana-pantilie committed:
Add simplification rule; modify tests outputs

ana-pantilie committed:
Add lookup simplification rules; update tests

ana-pantilie committed:
Remove old failing directory

ana-pantilie committed:
All Java tests; failing states; regenerate

ana-pantilie committed:
Fix assignment-18

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-26-spec.k.out:
```diff
@@ -0,0 +1,5 @@
+<k>
+  assignmentResult ( 0 |-> 5
+  1 |-> 4
+  X:MyId |-> 3 ) ~> .
+</k>
```
I think this gets simplified correctly.

On the Java backend, though, the expected result is:
```
<k>
  assignmentResult ( 1 |-> 4
  ( ( X |-> 3 ) [ 0 <- 5 ] ) )
</k>
```
It's also in the `failing` directory. So I don't know what to make of this test.

ana-pantilie committed:
Fix assigment-29

ana-pantilie committed:
Fix assignment-5

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/remove-6-spec.k.out:
```diff
@@ -0,0 +1,41 @@
+  } )
+#And
+  <k>
+    removeResult ( MAP [ Y:MyId <- undef ] [ Z:MyId ] ~> . ) ~> .
```
The expected Java result is:
```
<k>
  removeResult ( ( MAP
  X |-> 1 ) [ Y <- undef ] [ Z ] )
</k>
```
Since `Z =/=K X`, I think the Haskell backend result (`removeResult ( MAP [ Y:MyId <- undef ] [ Z:MyId ] )`) is correct.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/remove-9-spec.k.out:
```diff
@@ -0,0 +1,47 @@
+  } )
+#And
+  <k>
+    removeResult ( MAP [ Y:MyId <- undef ] [ Z:MyId ] ~> . ) ~> .
```
I think this is similar to `remove-6`.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-13-spec.k.out:
```diff
@@ -0,0 +1,24 @@
+  #Not ( {
```
This does not fail on the Java backend.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-20-spec.k.out:
```diff
@@ -0,0 +1,24 @@
+  #Not ( {
```
This does not fail on the Java backend.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-11-spec.k.out:
```diff
@@ -0,0 +1,51 @@
+  #Not ( {
```
This fails on the Java backend as well.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-12-spec.k.out:
```diff
@@ -0,0 +1,24 @@
+  #Not ( {
```
This fails on the Java backend as well.

ana-pantilie replied:
Does not fail on Haskell backend after 7735052.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-10-spec.k.out:
```diff
@@ -0,0 +1,37 @@
+  #Not ( {
```
If I'm not mistaken, we already discussed that we can't support this at the moment, so it will go to `failing`.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-11-spec.k.out:
```diff
@@ -0,0 +1,38 @@
+  #Not ( {
```
If I'm not mistaken, we already discussed that we can't support this at the moment, so it will go to `failing`.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-6-spec.k.out:
```diff
@@ -0,0 +1,23 @@
+  #Not ( {
```
If I'm not mistaken, we already discussed that we can't support this at the moment, so it will go to `failing`.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup-5-spec.k.out:
```diff
@@ -0,0 +1,38 @@
+  #Not ( {
```
This does not fail on the Java backend.

ana-pantilie replied:
Outdated after 7735052.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup-6-spec.k.out:
```diff
@@ -0,0 +1,42 @@
+  #Not ( {
```
This does not fail on the Java backend.

ana-pantilie replied:
Outdated after 7735052.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup-7-spec.k.out:
```diff
@@ -0,0 +1,44 @@
+  #Not ( {
```
This does not fail on the Java backend.

ana-pantilie replied:
Outdated after 7735052.

ana-pantilie committed:
Add new update rule, remove rule, regenerate tests

ana-pantilie commented on k-distribution/include/kframework/builtin/domains.md:
```diff
@@ -407,9 +407,34 @@ endmodule
+   rule M:Map [ K <- V ] => (K |-> V M) requires notBool (K in_keys(M)) [simplification]
+   rule M:Map [ K <- _ ] [ K <- V ] => M [ K <- V ] [simplification]
+   // rule M:Map [ K1 <- V1 ] [ K2 <- V2 ] => M [ K2 <- V2 ] [ K1 <- V1 ] requires K1 =/=K K2 andBool K2 in_keys(M) [simplification]
+   rule (K1 |-> V1 M:Map) [ K2 <- V2 ] => (K1 |-> V1 (M [ K2 <- V2 ])) requires K1 =/=K K2 [simplification]
```
After adding the rule at 419 (in 7735052), `assignment-15` would loop if I kept 418 as well.

ana-pantilie committed:
Remove accidentally added archive

ana-pantilie committed:
Fix assignment-10 output

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-10-spec.k.out:
```diff
@@ -0,0 +1,28 @@
+    Z
+  } )
+#And
+  <k>
+    assignmentResult ( ( MAP [ Y:MyId <- 2 ]
+    X:MyId |-> 1 ) [ Z:MyId <- 3 ] ) ~> .
+  </k>
```
This is consistent with the Java backend result, meaning that `[ Y <- 2]` appears before `X |-> 1`. I believe this is a consequence of adding the simplification rule from 419 in `domains.k` (7735052). There are other tests with the same behavior. ~I believe that this rule application would only be visible (checked) in the test if we kept the RHS as `fail`.~ Edit: we decided on using `.` on the RHS for all the claims.

ttuegel commented on k-distribution/include/kframework/builtin/domains.md:
```diff
@@ -407,9 +407,34 @@ endmodule
+   rule (K |-> _ M:Map) [ K <- V ] => (K |-> V M) [simplification]
+   rule M:Map [ K <- V ] => (K |-> V M) requires notBool (K in_keys(M)) [simplification]
+   rule M:Map [ K <- _ ] [ K <- V ] => M [ K <- V ] [simplification]
+   // rule M:Map [ K1 <- V1 ] [ K2 <- V2 ] => M [ K2 <- V2 ] [ K1 <- V1 ] requires K1 =/=K K2 andBool K2 in_keys(M) [simplification]
```
If this passes all the tests, I think we can go ahead and delete this rule.

ana-pantilie committed:
Revert tests to use . on RHS; regenerate

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-29-spec.k.out:
```diff
@@ -1 +1,23 @@
+  <k>
+    assignmentResult ( MAP
+    X:MyId |-> BAL_TO:Int
+    Y:MyId |-> BAL_FROM +Int -30 ) ~> .
```
Should this be `-Int 30` instead?

ttuegel replied:
I think there is a simplification rule for this, but maybe I'm thinking of EVM?

ana-pantilie replied:
No, the arithmetic simplification rules are in `domains.k` now. But it looks like this particular result comes from the application of [this](https://github.com/kframework/k/blame/master/k-distribution/include/kframework/builtin/domains.md#L1028) rule.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-30-spec.k.out:
```diff
@@ -1 +1,23 @@
+  <k>
+    assignmentResult ( MAP
+    foo ( X ) |-> BAL_TO:Int
+    foo ( Y ) |-> BAL_FROM +Int -30 ) ~> .
```
Should this be `-Int 30` instead?

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/assignment-31-spec.k.out:
```diff
@@ -1 +1,23 @@
+  <k>
+    assignmentResult ( MAP
+    foo ( X ) |-> BAL_TO:Int
+    foo ( Y ) |-> BAL_FROM +Int -30 ) ~> .
```
Should this be `-Int 30` instead?

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-2-spec.k.out:
```diff
@@ -1 +1,9 @@
-#Top
+  <k>
+    inkeysResult ( Y:MyId in_keys ( M ) ) ~> .
```
Why doesn't this get simplified further?

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/inkeys-3-spec.k.out:
```diff
@@ -1 +1,9 @@
-#Top
+  <k>
+    inkeysResult ( Y:MyId in_keys ( M ) ) ~> .
```
Why doesn't this get simplified further?

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup-23-spec.k.out:
```diff
@@ -1 +1,16 @@
+  } )
+#And
+  <k>
+    lookupResult ( ( X:MyId |-> 3 ) [ x <- 5 ] [ X:MyId ] ~> . ) ~> .
```
~Why doesn't this get simplified any further?~ I think this is correct, actually.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup-9-spec.k.out:
```diff
@@ -1 +1,47 @@
+  } )
+#And
+  <k>
+    lookupResult ( MAP [ Z:MyId ] ~> . ) ~> .
```
I think this is actually the correct result, since `Y =/=K Z`.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/remove-17-spec.k.out:
```diff
@@ -1 +1,28 @@
+  } )
+#And
+  <k>
+    removeResult ( ( X:MyId |-> 3 ) [ x <- undef ] [ X:MyId ] ~> . ) ~> .
```
I think this is actually the correct result, since `X =/=K y`.

ana-pantilie commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/remove-2-spec.k:
```diff
@@ -6,6 +6,6 @@ requires "map-tests.k"
     imports MAP-TESTS
 
-    claim <k> remove ( (MAP:Map X:MyId |-> 1) [Y:MyId <- undef] [ Z:MyId ] ) => removeResult ( MAP [ Z ] ) </k> requires Y ==K X andBool Z  ==K Y andBool Z  ==K X
+    claim <k> remove ( (MAP:Map X:MyId |-> 1) [Y:MyId <- undef] [ Z:MyId ] ) => . </k> requires Y ==K X andBool Z  ==K Y andBool Z  ==K X
```
The output is still `#Top` here, and I'm not sure if this is right.

ttuegel replied:
That is correct. All the keys are the same. Looking up the key after deleting it is `\bottom`.

ana-pantilie committed:
Clean-up

ana-pantilie committed:
Remove inkeys2 inkeys3, regenerate

ttuegel commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup-23-spec.k.out:
```diff
@@ -1 +1,16 @@
+  {
+    true
+  #Equals
+    X:MyId in_keys ( ( X:MyId |-> 3 ) [ x <- 5 ]
```
I think the result above is correct, but this should be simplified further. The `X` is obviously in the keys of the map.

ttuegel commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup-9-spec.k.out:
```diff
@@ -1 +1,47 @@
+    X:MyId in_keys ( MAP )
+  }
+#And
+  {
+    true
+  #Equals
+    Z:MyId in_keys ( ( MAP
+    X:MyId |-> 1 ) [ Y:MyId <- 2 ] )
+  }
+#And
+  {
+    true
+  #Equals
+    Z:MyId in_keys ( MAP )
+  }
+#And
+  {
+    true
+  #Equals
+    Z:MyId in_keys ( MAP [ Y:MyId <- 2 ]
+    ( X:MyId |-> 1 ) )
+  }
```
This is a little weird. Strictly, we only need the `{ true #Equals Z:MyId in_keys ( MAP ) }` side condition.

ttuegel commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/remove-17-spec.k.out:
```diff
@@ -1 +1,28 @@
+    removeResult ( ( X:MyId |-> 3 ) [ x <- undef ] [ X:MyId ] ~> . ) ~> .
+  </k>
+#And
+  {
+    false
+  #Equals
+    y in_keys ( ( X:MyId |-> 3 ) [ x <- undef ] )
+  }
```
The result above is correct, but I think this should be simplified away.

ttuegel commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/remove-17-spec.k.out:
```diff
@@ -1 +1,28 @@
+    y in_keys ( ( X:MyId |-> 3 ) [ x <- undef ] )
+  }
+#And
+  {
+    true
+  #Equals
+    X:MyId in_keys ( ( X:MyId |-> 3 ) [ x <- undef ]
+    ( y |-> 4 ) )
+  }
```
This should also be simplified, I think.

ttuegel commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/remove-17-spec.k.out:
```diff
@@ -1 +1,28 @@
+    ( y |-> 4 ) )
+  }
+#And
+  {
+    true
+  #Equals
+    X:MyId in_keys ( ( X:MyId |-> 3 ) [ x <- undef ] )
+  }
```
I think this should _not_ be simplified. We don't know that `X =/= x`.

ttuegel committed:
Merge branch 'master' into haskell-symbolic-lookup

ttuegel commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/lookup:
```diff
@@ -0,0 +1,24 @@
+claim <k> lookup ( (X:MyId |-> 1 MAP:Map Y:MyId |-> 2) [ Y:MyId ] ) => . </k> requires X =/=K Y
+claim <k> lookup ( (Y:MyId |-> 1 MAP:Map X:MyId |-> 2) [ Y:MyId ] ) => . </k> requires X =/=K Y
+claim <k> lookup ( (X:MyId |-> 1 Y:MyId |-> 2 MAP:Map) [ Y:MyId ] ) => . </k> requires X =/=K Y
+claim <k> lookup ( (Y:MyId |-> 1 X:MyId |-> 2 MAP:Map) [ Y:MyId ] ) => . </k> requires X =/=K Y
```
Could you please add

```.k
claim <k> lookup ( (Y:MyId |-> 1 X:MyId |-> 2) [ Y:MyId ] ) => . </k>
```

I'm curious to see what the backend does here.

ana-pantilie replied:
The result is:
```
<k>
  lookupResult ( 1 ~> . ) ~> .
</k>
```

ttuegel commented on k-distribution/include/kframework/builtin/domains.md:
```diff
@@ -408,9 +408,33 @@ endmodule
+  rule (K1 |-> V1 M:Map) [ K2 <- undef ] => (K1 |-> V1 (M [ K2 <- undef ])) requires K1 =/=K K2 [simplification]
 
+  // Symbolic lookup
+  rule (K |-> V _:Map) [ K ] => V [simplification]
```
I have a pointed question about this rule, below.

ttuegel commented on k-distribution/tests/regression-new/map-symbolic-tests-haskell/remove:
```diff
@@ -0,0 +1,17 @@
+claim <k> remove ( (MAP:Map X:MyId |-> 1) [Y:MyId <- undef] [ Z:MyId ] ) => . </k> requires Z =/=K Y andBool Z =/=K X
+claim <k> remove ( (MAP:Map X:MyId |-> 1) [Y:MyId <- undef] [ Z:MyId ] ) => . </k> requires Z =/=K Y andBool Z  ==K X andBool Y =/=K X
+claim <k> remove ( (X:MyId |-> 3 y |-> 4) [ x <- undef ] [ X ] ) => . </k> requires X =/=K x
+claim <k> remove ( (X:MyId |-> 3 y |-> 4) [ x <- undef ] [ X ] ) => . </k> requires X =/=K y
```
Could you please add

```
claim <k> remove ( (X:MyId |-> 1 Y:MyId |-> 2) [ X <- undef ] ) => . </k>
```

ana-pantilie replied:
The result is:
```
<k>
  removeResult ( Y:MyId |-> 2 ~> . ) ~> .
</k>
```

ttuegel commented on k-distribution/include/kframework/builtin/domains.md:
```diff
@@ -408,9 +408,33 @@ endmodule
 
-   rule #Ceil(@M:Map [@K:KItem]) => {(@K in_keys(@M)) #Equals true} #And #Ceil(@M) #And #Ceil(@K) [anywhere, simplification]
+  // Symbolic remove
+  rule (K |-> _ M:Map) [ K <- undef ] => M [simplification]
```
I also have a pointed question about this rule below.

ana-pantilie committed:
Review: add two new tests

ttuegel reviewed (approved):
This looks good to me! @dwightguth could you have a look?

ana-pantilie committed:
Merge branch 'master' into haskell-symbolic-lookup

ana-pantilie committed:
Rebuild

ana-pantilie committed:
Merge branch 'haskell-symbolic-lookup' of github.com:ana-pantilie/k into haskell-symbolic-lookup

ana-pantilie committed:
Merge branch 'master' into haskell-symbolic-lookup

ana-pantilie committed:
Remove unused fail constructor

ana-pantilie commented:
My understanding is that the two Nix tests are not required for merging, and should be fixed after the holidays.

This should be ready for review, but the main test pipeline has been experiencing intermittent failures which I haven't been able to reproduce locally. The error is always `java.lang.OutOfMemoryError: GC overhead limit exceeded`, and it only started appearing after merging master in a3350e3.

The last time I merged master the test result was: https://office.runtimeverification.com/jenkins/blue/organizations/jenkins/k/detail/PR-1661/24/pipeline. I noticed that an additional error appeared: `org.kframework.kore.convertors.KILTransformation$VisitingException`, something related to `map-tests.k`. After making a small change to that file (removing a constructor I wasn't using anyway and forgot to remove) the tests passed again.

@dwightguth Do you think removing that constructor should have any effect on what's happening here, or is it just a coincidence? And have you ever seen this sort of error before/have any ideas what I could do to debug it?

dwightguth reviewed (approved):
lgtm. Feel free to merge from my perspective if @ttuegel is happy.

ana-pantilie committed:
Merge branch 'master' into haskell-symbolic-lookup

ttuegel committed:
Trigger workflows

ttuegel committed:
Merge branch 'master' into haskell-symbolic-lookup
